core/src/main/java/org/apache/iceberg/rest/responses/FetchPlanningResultResponseParser.java

```diff
@@ -0,0 +1,138 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.rest.responses;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.io.IOException;
+import java.util.List;
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.rest.PlanStatus;
+import org.apache.iceberg.rest.RESTContentFileParser;
+import org.apache.iceberg.rest.RESTFileScanTaskParser;
+import org.apache.iceberg.util.JsonUtil;
+
+public class FetchPlanningResultResponseParser {
+  private static final String PLAN_STATUS = "plan-status";
+  private static final String PLAN_TASKS = "plan-tasks";
+  private static final String FILE_SCAN_TASKS = "file-scan-tasks";
+  private static final String DELETE_FILES = "delete-files";
+
+  private FetchPlanningResultResponseParser() {}
+
+  public static String toJson(FetchPlanningResultResponse response) {
+    return toJson(response, false);
+  }
+
+  public static String toJson(FetchPlanningResultResponse response, boolean pretty) {
+    return JsonUtil.generate(gen -> toJson(response, gen), pretty);
+  }
+
+  public static void toJson(FetchPlanningResultResponse response, JsonGenerator gen)
+      throws IOException {
+    Preconditions.checkArgument(null != response, "Invalid response: planTableScanResponse null");
+
+    gen.writeStartObject();
+    if (response.planStatus() != null) {
+      gen.writeStringField(PLAN_STATUS, response.planStatus().status());
+    }
+
+    if (response.planTasks() != null) {
+      gen.writeArrayFieldStart(PLAN_TASKS);
+      for (String planTask : response.planTasks()) {
+        gen.writeString(planTask);
+      }
+      gen.writeEndArray();
+    }
+
+    List<DeleteFile> deleteFiles = null;
+    if (response.deleteFiles() != null) {
+      deleteFiles = response.deleteFiles();
+      gen.writeArrayFieldStart(DELETE_FILES);
+      for (DeleteFile deleteFile : deleteFiles) {
+        RESTContentFileParser.toJson(deleteFile);
+      }
+      gen.writeEndArray();
+    }
+
+    if (response.fileScanTasks() != null) {
+      gen.writeArrayFieldStart(FILE_SCAN_TASKS);
+      for (FileScanTask fileScanTask : response.fileScanTasks()) {
+        RESTFileScanTaskParser.toJson(fileScanTask, deleteFiles, gen);
```

Comment:
See my comment below, passing the deleteFIles like this is incorrect. We should be passing only the indices of the top level list that are relevant for the task we are serializing.

Reply:
```
    Map<String, Integer> deleteFilePathToIndex = Maps.newHashMap();
    if (response.deleteFiles() != null) {
      deleteFiles = response.deleteFiles();
      gen.writeArrayFieldStart(DELETE_FILES);
      for (int i = 0; i < deleteFiles.size(); i++) {
        deleteFilePathToIndex.put(deleteFiles.get(i).location(), i);
        RESTContentFileParser.toJson(deleteFiles.get(i), gen);
      }
      gen.writeEndArray();
    }

    if (response.fileScanTasks() != null) {
      List<Integer> deleteFileReferences = Lists.newArrayList();
      gen.writeArrayFieldStart(FILE_SCAN_TASKS);
      for (FileScanTask fileScanTask : response.fileScanTasks()) {
        if (deleteFiles != null) {
          Set<String> taskDeletes =
              fileScanTask.deletes().stream()
                  .map(delete -> delete.location())
                  .collect(Collectors.toSet());
          for (String taskDelete : taskDeletes) {
            deleteFileReferences.add(deleteFilePathToIndex.get(taskDelete));
          }
        }

        RESTFileScanTaskParser.toJson(fileScanTask, deleteFileReferences, gen);
      }
```

Reply:
Thanks @amogh-jahagirdar for the clarification here.

Reply:
I was looking at this again locally, correction in the second half:

```

    if (response.fileScanTasks() != null) {
      gen.writeArrayFieldStart(FILE_SCAN_TASKS);
      for (FileScanTask fileScanTask : response.fileScanTasks()) {
      // the list should be initialized here, the delete file references are per file scan task
      List<Integer> deleteFileReferences = Lists.newArrayList();

        if (deleteFiles != null) {
          for (DeleteFile taskDelete : taskDeletes) {
            deleteFileReferences.add(deleteFilePathToIndex.get(taskDelete.location()));
          }
        }

        RESTFileScanTaskParser.toJson(fileScanTask, deleteFileReferences, gen);
      }
```

Anyways, with tests this will all be made clear

Reply:
Thanks @amogh-jahagirdar